### PR TITLE
Added Ember.computed.uniqBy

### PR DESCRIPTION
**Pull request by seanjohnson08**
Almost all of the enumerable methods that serve to produce some new result from a list follow the following pattern:

| Enumerate by object | Enumerate by object property |
| --- | --- |
| `filter` | `filterBy` |
| `map` | `mapBy` |
| `find` | `findBy` |
| `reject` | `rejectBy` |
| `uniq` | ???? |

Except `uniq`! My proposal is to add a `uniqBy` method to both the `Enumerable` class and to the computed reduce macros to allow the user to reduce a list down to its unique values given some key. A prime example is if I'm dealing with a list of records, or objects returned that have a primary key:

`var records = Ember.A([
{id: 1, name: 'Adam' },
{id: 2, name: 'Eve' },
{id: 1, name: 'Adam'}
]);`

If I wanted to remove the duplicate records, I would then be able to use the result of `records.uniqBy('id')` to get a list of just the unique records in the list.

This would not alter or modify the existing `uniq` methods, it would just serve to fill in a blank.

**Note**: The code I've pull requested does not actually add `Enumerable.uniqBy`, only `computed.uniqBy`. This PR is just to get some feedback. If the feedback is positive, I'll implement the `Enumerable` method as well.

### CODE REVIEW

**Comment by mixonic** [comment]
@seanjohnson08 I'll bound this off some other core folks. My personal inclination is to land it. For that to happen (per [CONTRIBUTING.md](https://github.com/emberjs/ember.js/blob/master/CONTRIBUTING.md)):

* This PR comment must be prefixed by `[FEATURE computed-uniqBy]`
* A feature flag such as `computed-uniqBy` must be added around the feature, and documented in `FEATURES.md` and `features.json`.
* Your patch does not expose `uniqBy` as a property on `computed` (am I mistaken?) That should also be done.

Thanks!

**Comment by stefanpenner** [comment]
seems like a good addon.

File: packages/ember-runtime/lib/computed/reduce_computed_macros.js [hunk 1]
```diff
@@ -362,6 +362,53 @@ export function uniq(...args) {
...
+  @method uniqBy
+  @for Ember.computed
+  @param {String} dependentKey
+  @param {String} propertyKey
+  @return {Ember.ComputedProperty} computes a new array with all the
+  unique elements from the dependent array
+  @public
+*/
+export function uniqBy(dependentKey, propertyKey) {
+  var seen = {};
```
Comment (stefanpenner) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: I believe this implementation is incorrect, given multiple instances the seen object is shared.

File: packages/ember-runtime/lib/computed/reduce_computed_macros.js [hunk 2]
```diff
@@ -362,6 +362,53 @@ export function uniq(...args) {
...
+
+  var callback = function(item) {
+    var value = get(item, propertyKey);
```
Comment (stefanpenner) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: i think this should be the guid of the result, not the result

Comment (stefanpenner) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: @rwjblue i should merge my weakmap stuff, it could be used here.

**Comment by seanjohnson08** [comment]
I'll do the work and get it cleaned up/ready to merge (and add associated documentation) if the computed moratorium has been lifted/the core team agrees this is a good addition.

**Comment by stefanpenner** [comment]
In today's meeting we agreed (do to the symmetry argument above) that this macro should be added.

File: packages/ember-runtime/lib/computed/reduce_computed_macros.js (hunk 2 above)
Comment (rwjblue) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: Yes, agreed. I believe it also fixes a couple bugs (with sortBy IIRC)...

**Comment by seanjohnson08** [comment]
Will open a new pull request once I address all of the above issues. I closed this because it ended up being too difficult to rebase on 2 months of Ember changes. :)

**Pull request closed without merging**